app/openai_ops.py

```diff
@@ -359,76 +359,98 @@ def context_length(
         raise NotImplementedError(error)
 
 
-# Adapted from https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb
+def get_encoding_for_model(model: str):
+    try:
+        return tiktoken.encoding_for_model(model)
+    except KeyError:
+        return tiktoken.get_encoding("cl100k_base")
+
+
+def get_tokens_per_message(model: str):
+    model_specific_tokens = {
+        GPT_3_5_TURBO_0613_MODEL: (3, 1),
+        GPT_3_5_TURBO_16K_0613_MODEL: (3, 1),
+        GPT_3_5_TURBO_1106_MODEL: (3, 1),
+        GPT_3_5_TURBO_0125_MODEL: (3, 1),
+        GPT_4_0314_MODEL: (3, 1),
+        GPT_4_32K_0314_MODEL: (3, 1),
+        GPT_4_0613_MODEL: (3, 1),
+        GPT_4_32K_0613_MODEL: (3, 1),
+        GPT_4_1106_PREVIEW_MODEL: (3, 1),
+        GPT_4_0125_PREVIEW_MODEL: (3, 1),
+        GPT_4_TURBO_PREVIEW_MODEL: (3, 1),
+        GPT_4_TURBO_2024_04_09_MODEL: (3, 1),
+        GPT_4O_2024_05_13_MODEL: (3, 1),
+        GPT_3_5_TURBO_0301_MODEL: (4, -1),
+    }
+    return model_specific_tokens.get(model, None)
+
+
+def handle_fallback_models(model: str, messages: List[Dict[str, Union[str, Dict[str, str]]]]):
+    model_fallbacks = {
+        GPT_3_5_TURBO_MODEL: GPT_3_5_TURBO_0125_MODEL,
+        GPT_3_5_TURBO_16K_MODEL: GPT_3_5_TURBO_16K_0613_MODEL,
+        GPT_4_MODEL: GPT_4_0613_MODEL,
+        GPT_4_TURBO_MODEL: GPT_4_TURBO_2024_04_09_MODEL,
+        GPT_4_32K_MODEL: GPT_4_32K_0613_MODEL,
+        GPT_4O_MODEL: GPT_4O_2024_05_13_MODEL,
+    }
+    if model in model_fallbacks:
+        return calculate_num_tokens(messages, model=model_fallbacks[model])
+    return None
+
+
+def raise_not_supported_error(model: str):
+    error = (
+        f"Calculating the number of tokens for model {model} is not yet supported. "
+        "See https://github.com/openai/openai-python/blob/main/chatml.md "
+        "for information on how messages are converted to tokens."
+    )
+    raise NotImplementedError(error)
+
+
+def encode_and_count_tokens(value, encoding):
+    if isinstance(value, str):
+        return len(encoding.encode(value))
+    elif isinstance(value, list):
+        return sum(encode_and_count_tokens(item, encoding) for item in value)
+    elif isinstance(value, dict):
+        return sum(encode_and_count_tokens(v, encoding) for k, v in value.items() if k != 'image_url')
+    return 0
+
+
 def calculate_num_tokens(
     messages: List[Dict[str, Union[str, Dict[str, str]]]],
-    model: str = GPT_3_5_TURBO_0613_MODEL,
+    model: str = GPT_3_5_TURBO_0613_MODEL
 ) -> int:
-    """Returns the number of tokens used by a list of messages."""
```

Comment:
I prefer less changes here. you can extract common parts for supporting new patterns but there is no need to completely rewrite.

Reply:
ok, just use ai to help refactor, I think it's too long